File: robotics_backend_ws/src/driver_wrapper_cpp_pkg/src/motor_ros_wrapper_withOOP.cpp
#include <ros/ros.h>
#include "../include/motor_driver.hpp"

#include <memory>
#include <map>
#include <vector>

#include <std_msgs/Int32.h>
#include <std_srvs/Trigger.h>

#include <diagnostic_msgs/DiagnosticStatus.h>
#include <diagnostic_msgs/KeyValue.h>

class MotorDriverROSWrapper
{
private:
    std::unique_ptr<MotorDriver> motor_;
    ros::Subscriber speed_command_subscriber_;

    ros::Publisher current_speed_publisher_;
    ros::Publisher motor_status_publisher_;
    ros::Timer current_speed_timer_;
    ros::Timer motor_status_timer_;

    double publish_current_speed_frequency_;
    double publish_motor_status_frequency_;

public:
    MotorDriverROSWrapper(ros::NodeHandle *nh)
    {

        int max_speed;
        if (!ros::param::get("~max_speed", max_speed))
            max_speed = 8;

        if (!ros::param::get("~publish_current_speed_frequency", publish_current_speed frequency_))
            publish_current_speed_frequency_ = 5.0;

        if (!ros::param::get("~publish_motor_status_frequency", publish_motor_status frequency_))
            publish_motor_status_frequency_ = 1.0;
        
        motor_.reset(new MotorDriver(8));

        speed_command_subscriber = nh->subscribe("speed_command", 10, &MotorDriverROSWrapper::callbackSpeedCommand, this);

        stop_motor_server_ = nh->advertiseService("stop_motor", &MotorDriverROSWrapper::callbackStop, this);

        current_speed_publisher_ = nh->advertise<std_msgs::Int32>("current_speed", 10);
        motor_status_publisher_ = nh->advertise<diagnotic_msgs::DiagnosticStatus>("motor_status", 10);

        current_speed_timer_ = nh->createTimer(ros::Duration(1.0 / publish_current_speed_frequency_), &MotorDriverROSWrapper::publishCurrentSpeed, this);
        motor_status_timer_ = nh->createTimer(ros::Duration(1.0 / publish_motor_status_frequency_), &MotorDriverROSWrapper::publishMotorStatus, this);
    }

    void publishCurrentSpeed(const ros::TimerEvent &event)
    {
        std_msgs::Int32 msg;
        msg.data = motor_->getSpeed();
        current_speed_publisher_.publish(msg);
    }

    void publishMotorStatus(const ros::TimerEvent &event)
    {
        diagnostic_msgs::DiagnosticStatus msg;
        std::map<std::string, std::string> status = motor_->getStatus();
        std::vector<diagnostic_msgs::KeyValue> values;

        for (auto const &x : status)
        {
            diagnostic_msgs::KeyValue value;
            value.key = x.first;
            value.value = x.second;
            values.push_back(value);
        }

        msg.values = values;
        motor_status_publisher_.publish(msg);
    }

    void callbackSpeedCommand(const std_msgs::Int32 &msg)
    {
        motor_->setSpeed(msg.data);
    }

    bool callbackStop(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &res)
    {
        stop();
        res.success = true;
        res.message = "Successfully stopped motor";
        return true;
    }

    void stop()
    {
        motor_->stop();
    }
};

int main(int argc, char const *argv[])
{
    ros::init(argc, argv, "motor_driver");
    ros::NodeHandle nh;

    ros::AsyncSpinner spinner(4);
    spinner.start();

    MotorDriverROSWrapper motor_driver_wrapper(&nh);
    ROS_INFO("Motor driver is now started");

    ros::waitForShutdown();
    motor_driver_wrapper.stop();

    return 0;
}

File: robotics_backend_ws/src/driver_wrapper_python_pkg/scripts/motor_ros_wrapper_withOOP.py
#!/usr/bin/env python

import rospy
# from motor_driver import MotorDriver
from ..src.my_robot_driver import MotorDriver

from std_msgs import Int32
from std_srvs import Trigger

from diagnostic_msgs.msg import DiagnosticStatus
from diagnostic_msgs.msg import KeyValue

class MotorDriverROSWrapper:

    def __init__(self):
        # self.motor = MotorDriver(max_speed=8)

        max_speed = rospy.get_param("~max_speed", 8)
        self.motor = MotorDriver(max_speed=max_speed)

        rospy.Subsriber("speed_command", Int32, self.callback_speed_command)

        rospy.Service("stop_motor", Trigger, self.callback_stop)
        
        publish_current_speed_frequency = rospy.get_param("~publish_current_speed_frequency" , 5.0)
        publish_motor_status_frequency = rospy.get_param("~publish_motor_status_frequency", 1.0)

        self.current_speed_pub = rospy.Publisher("current_speed" , Int32 , queue_size = 10)
        self.motor_status_pub = rospy.Publisher("motor_status" , DiagnosticStatus , queue_size = 1)

        rospy.Timer(rospy.Duration(1.0/publish_current_speed_frequency) , self.publish_current_speed)
        rospy.Timer(rospy.Duration(1.0/publish_motor_status_frequency) , self.publish_motor_status)

    def publish_current_speed(self, event= None):
        self.current_speed_pub.publish(self.motor.get_speed())

    def publish_motor_status (self , event = None):
        status = self.motor.get_status()
        data_list = []

        for key in status:
            data_list.append(KeyValue(key , str(status[key])))

            msg = DiagnosticStatus()
            msg.values = data_list

            self.motor_status_pub.publish(msg)

    def stop(self):
        self.motor.stop()
    
    def callback_speed_command(self, msg):
        self.motor.set_speed(msg.data)


    def callback_stop(self, req):
        self.stop()
        return {"success": True, "message": "Motor has been stopped"}

if __name__ == "__main__":
    rospy.init_node("motor_driver")

    motor_driver_wrapper = MotorDriverROSWrapper()
    rospy.on_shutdown(motor_driver_wrapper.stop())

    rospy.loginfo("Motor driver is now started,ready to get commands")
    rospy.spin()
